### ISSUE TITLE
SDWebImage循环引用问题

### ISSUE THREAD

**Issue by zjutwhw** [rs/SDWebImage#2493]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

Info                    | Value                               |
-------------------------|-------------------------------------|
Platform Name           | e.g. ios / macos / tvos / watchos
Platform Version        | e.g. 11.0 / 10.13.0 / 11.0 / 4.0
SDWebImage Version      | e.g. 4.2.0 / 4.1.0
Integration Method      | e.g. carthage / cocoapods / manually
Xcode Version           | e.g. Xcode 9 / Xcode 8
Repro rate              | e.g. all the time (100%) / sometimes x% / only once
Repro with our demo prj | e.g. does it happen with our demo project?
Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

SDWebImage存在循环引用问题，使用不慎，可能会导致VC无法释放，review代码发现以下加载完URL对应的image之后，没有对id <SDWebImageOperation> operation进行释放，导致外部的completedBlock一直被强持有，如果在completedBlock中使用了self，就可能造成retain Cycles，具体如下：

UIView+WebCache.m  <100行>:

```
        id <SDWebImageOperation> operation = [manager loadImageWithURL:url options:options progress:combinedProgressBlock completed:^(UIImage *image, NSData *data, NSError *error, SDImageCacheType cacheType, BOOL finished, NSURL *imageURL) {
            __strong __typeof (wself) sself = wself;
            if (!sself) { return; }
#if SD_UIKIT
            [sself sd_removeActivityIndicator];
#endif
            **// 修复问题
            // release operation ,avoid 'completedBlock' retain Cycles 
            [sself sd_removeImageLoadOperationWithKey:validOperationKey];**

...
```
请作者尽快修复问题~
感谢开源！！

**Comment by dreampiggy** [comment]
This does not cause a retain cycle. Let me tell the reason.

First, for each `SDWebImageCombineOperation`, there are two object will retain it. The first one, it a `NSArray` in `SDWebImageManager.runningOperation`. Which hold a strong reference to all pending loading tasks, until it been finished or cancelled. This one should keep, or the operation will been dealloced.

Second one, it the `UIView+WebCacheOperation`, each view use a strong-weak `NSMapTable` in associated object (strong property). Because the maptable is a weak reference to the operation, after the `runnningOperation` array remove the operation, the opeartion will be dealloced.

Your question, maybe related to your business code. For example, if you want to immediatelly cancel the network loading after your VC been poped or dismissed, you should not capture `self` in the completionBlock. Instead, use a weak capture list for Swift, or Strong-Weak pair in Objective-C, to make the completionBlock does not capture your `self` can cause a temp retain cycle. Note even it's a retain cycle, after loading finished or cancel, all will be dealloced, is just a temp.

If you do not want to cancel the loading, for example, later you may request the same URL and you want to cache it and do some logic after loading, even if your VC disappeared. You should capture the `self` in the completion, to avoid operation been dealloced. It's also a possible use case, it does not means retain cycle is always bad, it's important to keep instance alive for some reason.

If you still have questions, see [Common-Problems Handle self capture in completion block](https://github.com/rs/SDWebImage/wiki/Common-Problems#handle-self-capture-in-completion-block)

**Comment by zjutwhw** [comment]
> ```ruby
> - (void)safelyRemoveOperationFromRunning:(nullable SDWebImageCombinedOperation*)operation {
>     @synchronized (self.runningOperations) {
>         if (operation) {
>             [self.runningOperations removeObject:operation];
>         }
>     }
> }
> ```
> 
> This method release the `id <SDWebImageOperation> operation`. @zjutwhw

for each SDWebImageCombineOperation, there are two object will retain it.

This problem has been fixed in the version 4.2.3

**Comment by zjutwhw** [comment]
> This does not cause a retain cycle. Let me tell the reason.
> 
> First, for each `SDWebImageCombineOperation`, there are two object will retain it. The first one, it a `NSArray` in `SDWebImageManager.runningOperation`. Which hold a strong reference to all pending loading tasks, until it been finished or cancelled. This one should keep, or the operation will been dealloced.
> 
> Second one, it the `UIView+WebCacheOperation`, each view use a strong-weak `NSMapTable` in associated object (strong property). Because the maptable is a weak reference to the operation, after the `runnningOperation` array remove the operation, the opeartion will be dealloced.
> 
> Your question, maybe related to your business code. For example, if you want to immediatelly cancel the network loading after your VC been poped or dismissed, you should not capture `self` in the completionBlock. Instead, use a weak capture list for Swift, or Strong-Weak pair in Objective-C, to make the completionBlock does not capture your `self` to cause a retain cycle. Note even it's a retain cycle, after loading finished or cancel, all will be dealloced, is just a temporary state.
> 
> But however, sometimes you need to keep `self` alive, for example, later you may request the same URL, so you want to cache it and do some logic after loading, even if your VC disappeared. You should capture the `self` in the completion, to avoid your businese code which may call `[self xxMethod]` fail. It's also a possible use case, it does not means retain cycle is always bad, it's important to keep instance alive for some reason.
> 
> If you still have questions, see [Common Problems: Handle self capture in completion block](https://github.com/rs/SDWebImage/wiki/Common-Problems#handle-self-capture-in-completion-block)

OK, It's a very nice explanation,I upgrade  the version of SDWebImage, and fix it
